python/tests/tsutil.py

```diff
@@ -1175,6 +1175,143 @@ def algorithm_T(ts):
         left = right
 
 
+class QuintuplyLinkedTree:
+    def __init__(self, n, root_threshold=1):
+        self.root_threshold = root_threshold
+        self.parent = np.zeros(n + 1, dtype=np.int32) - 1
+        self.left_child = np.zeros(n + 1, dtype=np.int32) - 1
+        self.right_child = np.zeros(n + 1, dtype=np.int32) - 1
+        self.left_sib = np.zeros(n + 1, dtype=np.int32) - 1
+        self.right_sib = np.zeros(n + 1, dtype=np.int32) - 1
+        self.num_samples = np.zeros(n + 1, dtype=np.int32)
+
+    def __str__(self):
+        s = "id\tparent\tlchild\trchild\tlsib\trsib\tnsamp\n"
+        for j in range(len(self.parent)):
+            s += (
+                f"{j}\t{self.parent[j]}\t"
+                f"{self.left_child[j]}\t{self.right_child[j]}\t"
+                f"{self.left_sib[j]}\t{self.right_sib[j]}\t"
+                f"{self.num_samples[j]}\n"
+            )
+        return s
+
+    def roots(self):
+        roots = []
+        u = self.left_child[-1]
+        while u != -1:
+            roots.append(u)
+            u = self.right_sib[u]
+        return roots
+
+    def remove_branch(self, p, c):
+        lsib = self.left_sib[c]
+        rsib = self.right_sib[c]
+        if lsib == -1:
+            self.left_child[p] = rsib
+        else:
+            self.right_sib[lsib] = rsib
+        if rsib == -1:
+            self.right_child[p] = lsib
+        else:
+            self.left_sib[rsib] = lsib
+        self.parent[c] = -1
+        self.left_sib[c] = -1
+        self.right_sib[c] = -1
+
+    def insert_branch(self, p, c):
+        assert self.parent[c] == -1, "contradictory edges"
+        self.parent[c] = p
+        u = self.right_child[p]
+        if u == -1:
+            self.left_child[p] = c
+            self.left_sib[c] = -1
+            self.right_sib[c] = -1
+        else:
+            self.right_sib[u] = c
+            self.left_sib[c] = u
+            self.right_sib[c] = -1
+        self.right_child[p] = c
+
+    def is_root(self, u):
+        return self.num_samples[u] >= self.root_threshold
```

Comment:
I don't understand what's going on here?

Reply:
Oh, I see - you only use this method on path ends. So like this method is actually "would this be a root at this threshold if it didn't have a parent" not "is this a root"?

Reply:
Yes, good call. This should really be ``is_potential_root``.

This is based on the current code by the way - the root threshold logic has been in there for a while. See the [C code](https://github.com/tskit-dev/tskit/blob/e1d64e30eeee291214020354faa787485efeabb4/c/tskit/trees.c#L3982) for example. This version is much simpler because we're not making special cases for the branch setting on roots.